File: bindings/ruby/test/test_version.rb
# frozen_string_literal: true

# Version surface over the raw SWIG-generated Ruby binding.
#
# Ports bindings/ruby/test/test_ptiff.rb's test_runtime_version and
# test_runtime_matches_compile_time onto Ptiff::ptiff_runtime_version /
# Ptiff::ptiff_compile_time_version, plus the out-parameter forms
# (Ptiff::ptiff_runtime_version_out / ptiff_compile_time_version_out).
# The *_out forms needed the OUTPUT typemap added to typemaps.i (SWIGRUBY
# block, "int *major"/"int *minor"/"int *patch") to surface as extra return
# values via SWIG_Ruby_AppendOutput -- without it there was no way to pass a
# bare int* by reference from pure Ruby.
require 'minitest/autorun'
require 'ptiff'

class TestVersion < Minitest::Test
  def test_runtime_version_fields
    # Keep in sync with the project version (bindings/rust/Cargo.toml, CHANGELOG.md).
    # Do NOT hard-code a specific minor here and forget to bump it on release —
    # the cross-check below (runtime == compile_time) is version-agnostic.
    v = Ptiff::ptiff_runtime_version
    assert_equal 0, v.major
    assert_equal 3, v.minor
    assert_equal 0, v.patch
  end

  def test_runtime_matches_compile_time
    r = Ptiff::ptiff_runtime_version
    c = Ptiff::ptiff_compile_time_version
    assert_equal [r.major, r.minor, r.patch], [c.major, c.minor, c.patch]
  end

  def test_version_out_params
    r_maj, r_min, r_pat = Ptiff::ptiff_runtime_version_out
    c_maj, c_min, c_pat = Ptiff::ptiff_compile_time_version_out
    assert_equal [r_maj, r_min, r_pat], [c_maj, c_min, c_pat]

    v = Ptiff::ptiff_runtime_version
    assert_equal [r_maj, r_min, r_pat], [v.major, v.minor, v.patch]
  end
end

File: bindings/python/test/test_version.py
"""Version surface over the raw SWIG-generated Python binding.

Ports the version coverage of bindings/python/test/test_ptiff.py onto the
SWIG output: both the struct-return (ptiff_runtime_version /
ptiff_compile_time_version) and the out-parameter (*_out) forms of the C ABI.
The *_out forms need the OUTPUT typemap added to typemaps.i (SWIGPYTHON
block, "int *major"/"int *minor"/"int *patch") -- without it, a plain int*
C ABI parameter has no pure-Python counterpart to pass by reference.
"""

import unittest

import ptiff


class TestVersion(unittest.TestCase):
    def test_runtime_version_matches_semver(self) -> None:
        v = ptiff.ptiff_runtime_version()
        s = f"{v.major}.{v.minor}.{v.patch}"
        self.assertRegex(s, r"^\d+\.\d+\.\d+$")

    def test_runtime_matches_compile_time(self) -> None:
        r = ptiff.ptiff_runtime_version()
        c = ptiff.ptiff_compile_time_version()
        self.assertEqual((r.major, r.minor, r.patch), (c.major, c.minor, c.patch))

    def test_version_out_params(self) -> None:
        rMaj, rMin, rPat = ptiff.ptiff_runtime_version_out()
        cMaj, cMin, cPat = ptiff.ptiff_compile_time_version_out()
        self.assertEqual((rMaj, rMin, rPat), (cMaj, cMin, cPat))

        v = ptiff.ptiff_runtime_version()
        self.assertEqual((rMaj, rMin, rPat), (v.major, v.minor, v.patch))


if __name__ == "__main__":
    unittest.main()
